samples/api/swapchain_recreation/swapchain_recreation.cpp

```diff
@@ -0,0 +1,929 @@
+}
+
+VkPhysicalDevice SwapchainRecreation::get_gpu_handle()
+{
+	return get_device().get_gpu().get_handle();
+}
+
+VkDevice SwapchainRecreation::get_device_handle()
+{
+	return get_device().get_handle();
+}
+
+SwapchainRecreation::SwapchainRecreation()
+{
+}
+
+SwapchainRecreation::~SwapchainRecreation()
+{
+	// Wait for device to be idle and clean up everything.
+	vkDeviceWaitIdle(get_device_handle());
+
+	for (PerFrame &frame : submit_history)
+	{
+		recycle_fence(frame.submit_fence);
+		recycle_semaphore(frame.acquire_semaphore);
+		vkDestroyCommandPool(get_device_handle(), frame.command_pool, nullptr);
+		for (SwapchainObjects &garbage : frame.swapchain_garbage)
+		{
+			cleanup_swapchain_objects(garbage);
+		}
+		frame.swapchain_garbage.clear();
+		assert(frame.present_semaphore == VK_NULL_HANDLE);
+	}
+
+	for (PresentOperationInfo &present_info : present_history)
+	{
+		if (present_info.cleanup_fence != VK_NULL_HANDLE)
+		{
+			vkWaitForFences(get_device_handle(), 1, &present_info.cleanup_fence, true, UINT64_MAX);
+		}
+		cleanup_present_info(present_info);
+	}
+
+	LOGI("Old swapchain count at destruction: {}", old_swapchains.size());
+
+	for (SwapchainCleanupData &old_swapchain : old_swapchains)
+	{
+		cleanup_old_swapchain(old_swapchain);
+	}
+
+	LOGI("Semaphore pool size at destruction: {}", semaphore_pool.size());
+	LOGI("Fence pool size at destruction: {}", fence_pool.size());
+
+	for (VkSemaphore semaphore : semaphore_pool)
+	{
+		vkDestroySemaphore(get_device_handle(), semaphore, nullptr);
+	}
+
+	for (VkFence fence : fence_pool)
+	{
+		vkDestroyFence(get_device_handle(), fence, nullptr);
+	}
+
+	cleanup_swapchain_objects(swapchain_objects);
+	if (swapchain != VK_NULL_HANDLE)
+	{
+		vkDestroySwapchainKHR(get_device_handle(), swapchain, nullptr);
+	}
+
+	if (render_pass != VK_NULL_HANDLE)
+	{
+		vkDestroyRenderPass(get_device_handle(), render_pass, nullptr);
+	}
+}
+
+bool SwapchainRecreation::prepare(vkb::Platform &platform)
+{
+	if (!VulkanSample::prepare(platform))
+	{
+		return false;
+	}
+
+	LOGI("USAGE:");
+	LOGI(" - Press v to enable v-sync");
+	LOGI(" - Press n to disable v-sync");
+
+	return true;
+}
+
+void SwapchainRecreation::create_render_context(vkb::Platform &platform)
+{
+	get_queue();
+	query_surface_format();
+	create_render_pass();
+	init_swapchain();
+}
+
+void SwapchainRecreation::prepare_render_context()
+{
+	// Nothing to do
+}
+
+void SwapchainRecreation::update(float delta_time)
+{
+	fps_timer += delta_time;
+	if (fps_timer > 1.0f)
+	{
+		LOGI("FPS: {}", static_cast<float>(frame_number - fps_last_logged_frame_number) / fps_timer);
+		fps_timer -= 1.0f;
+		fps_last_logged_frame_number = frame_number;
+	}
+
+	++frame_number;
+
+	setup_frame();
+
+	if (desired_present_mode != current_present_mode)
+	{
+		recreate_swapchain();
+	}
+
+	uint32_t index;
+	auto     res = acquire_next_image(&index);
+
+	// Handle outdated error in acquire.
+	if (res == VK_SUBOPTIMAL_KHR || res == VK_ERROR_OUT_OF_DATE_KHR)
```

Comment:
Hi @ShabbyX, I'm wondering if the logic here might be problematic on Android. Specifically, if pre-rotation is not honoured[1] the WSI calls will tend to return `VK_SUBOPTIMAL_KHR` since Android 10 IIRC. If we imagine that happening here, it seems like the below branch would be entered, `recreate_swapchain()` is called, however then it would early-out because surface dimensions are unchanged, and then `acquire_next_image()` is called on the line below.

On the face of it, that seems like we'd have `acquire_next_image()` called twice in a row without any swap-chain recreation in between -- which seems incorrect.

It looks like the code as it is now effectively assumes VK_SUBOPTIMAL_KHR is an error code which should trigger re-creation (thus also a re-acquire), but (a) the current code seems like it won't actually do re-creation in this scenario (leading to a potential bug), and (b) many apps still do not honour pre-rotation, so treating VK_SUBOPTIMAL_KHR this way in the sample might be a bad idea (would lead to unnecessary swapchain re-creation _every frame_ in such apps if they try to implement similar code as in the sample, as `VK_SUBOPTIMAL_KHR` will keep being returned given the app never actually start pre-rotating)

I should say, I say this only having glancing over the code here so I might've missed something, but wanted to check if this could be a concern. :)

[1] Note that the sample as-is _does_ seem to honour it, as you have `info.preTransform     = surface_properties.currentTransform`. A way to confirm the above issue might be to change that and see if you now get SUBOPTIMAL returned every frame.

Reply:
Hi. You are right, the code _should_ be modified to account for rotation (like swap the dimensions after the query if 90/270 degree rotations). Probably the offset/extents of the clear command should also be rotated.

May I ask for help here though? I'm not really well setup to run the samples on Android, so I'd appreciate it if someone can follow up with the necessary changes to make it correct in the presence of rotation. I agree it's important to get that right in the sample, but it's rather beyond the scope of what I signed up to demonstrate in this sample.

Reply:
> It looks like the code as it is now effectively assumes VK_SUBOPTIMAL_KHR is an error code which should trigger re-creation

The thing is it's different per platform. Like on non-Android if you get SUBOPTIMAL because the user dragged the window corner and changed the size, you _do_ want to recreate the swapchain.

An alternative here is to just set the preTransform to 0, and comment that "rotation is not handled, please don't do it / you need to handle it properly in your application --> see this other sample (if we have any)".

Reply:
Aah, I see! That's a bit annoying in that case... Just a comment sounds pretty good to me, either keeping the code as-is but pointing out that if you modify / don't handle pre-rotation it might be a good idea to _also_ modify the re-creation-logic as well -- or like you say stop handling pre-rotation and point out the consequences.

I think my preference would be the former because then the code is good and would work fine on Android, but it still helpfully points out this is something to keep in mind if porting it elsewhere (specifically to somewhere where pre-rotation isn't handled).

Reply:
I opted for using identity for preTransform and ignore SUBOPTIMAL on Android + added warnings. Dealing with rotation is very important, but I feel like it deserves its own sample where it can be thoroughly explained.